### ISSUE TITLE
Rewrite encoders to consume pixels of each frame in an abstract way

### ISSUE THREAD

**Issue by dacap** [aseprite/aseprite#3008]
There are several issues that could be fixed adding a new way to export images scaling each frame "on the fly" when we are encoding the file format:

- [x] #1112
- [x] #1749
- [x] #2569
- [x] Maybe #2897
- [x] #3194
- [x] #3007
- [ ] #493

The basic idea is to have a new interface like (e.g.):

```c++
class IAbstractImage {
public:
  virtual ~IAbstractImage() { }
  virtual frame_t totalFrames() = 0;
  virtual void moveToFrame(frame_t fr) = 0;
  virtual const ImageSpec& imageSpec() = 0;
  virtual void getScanline(int y, base::buffer& buf) = 0;
};
```

And each encoder could read each scanlines without accessing the `doc::Image*` directly. So we don't have to resize the image on memory (which will allow us to encode huge resizing factors).

Some details:

* We could start implementing the new interface only for `.png` and   `.gif` files (as they are the most common formats)
* It can be implemented with a new file format flag similar to `FILE_SUPPORT_SEQUENCES`, where if the format uses the new flag, it means that will use the new `IAbstractImage` interface.
* `.gif` files require to generate a palette from each input frame/image, this input image can be the original one instead of the scaled one if the scaled one is bigger. If the scaled one is smaller   we could use the new one instead as it might contain fewer colors than the original image (so it's tricky and depends on the encoder   how they will use each image, the original vs. the scaled one).

Some work was already done but it's not yet public/available.

**Comment by MohamedAlaa201** [comment]
I would be happy to help/work on implementing this.

**Comment by dacap** [comment]
Hi @MohamedAlaa201, actually I've some work-in-progress in a stash, I'll try to push the work in a public branch in these days.

**Comment by dacap** [comment]
4aa5fedfec2b809619763eac68b95fca1e46f20c is a possible fix for #2569, #2897, #3194
